File: LeetCode/0480. Sliding Window Median/Solution.java
/**
 * 42 / 42 test cases passed.
 * Runtime: 45 ms
 * Memory Usage: 43.3 MB 
 */
class Solution {
    private double get(int[] nums) {
        int sz = nums.length;
        return (nums[(sz - 1) / 2] * 1.0 / 2) + (nums[sz / 2] * 1.0 / 2);
    }

    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    } 

    public double[] medianSlidingWindow(int[] nums, int k) {
        int[] slide = Arrays.copyOfRange(nums, 0, k);
        Arrays.sort(slide);

        int n = nums.length;
        double[] ans = new double[n - k + 1];
        ans[0] = get(slide);
        for (int i = 0; i < n - k; i++) {
            int index = Arrays.binarySearch(slide, nums[i]);
            slide[index] = nums[i + k];
            while (index < slide.length - 1 && slide[index] > slide[index + 1]) {
                swap(slide, index, index + 1);
                index++;
            }
            while (index > 0 && slide[index] < slide[index - 1]) {
                swap(slide, index, index - 1);
                index--;
            }
            ans[i + 1] = get(slide);
        }
        return ans;
    }
}

File: LeetCode/0480. Sliding Window Median/solution.cpp
/**
 * 42 / 42 test cases passed.
 * Runtime: 32 ms
 * Memory Usage: 14.1 MB 
 */
class Solution {
public:
    priority_queue<int> small;
    priority_queue<int, vector<int>, greater<int>> big;
    unordered_map<int, int> rec;

    double get(int &k) {
        if (k & 1) return small.top();
        return small.top() * 0.5 + big.top() * 0.5;
    }

    vector<double> medianSlidingWindow(vector<int>& nums, int k) {
        int n = nums.size();
        for (int i = 0; i < k; ++ i) small.push(nums[i]);
        for (int i = 0, k2 = k / 2; i < k2; ++ i) {
            big.push(small.top());
            small.pop();
        }

        vector<double> ans{ get(k) };
        for (int i = k; i < n; ++ i) {
            int balance = 0;  // size(small) - size(big)
            int left = nums[i - k];
            int right = nums[i];

            ++ rec[left];
            if (!small.empty() && left <= small.top()) {
                -- balance;
            } else {
                ++ balance;
            }
            if (!small.empty() && right <= small.top()) {
                ++ balance;
                small.push(right);
            } else {
                -- balance;
                big.push(right);
            }

            if (balance > 0) {
                big.push(small.top());
                small.pop();
            } else if (balance < 0) {
                small.push(big.top());
                big.pop();
            }

            while (!small.empty() && rec[small.top()] > 0) {
                -- rec[small.top()];
                small.pop();
            }
            while (!big.empty() && rec[big.top()] > 0) {
                -- rec[big.top()];
                big.pop();
            }
            ans.push_back(get(k));
        }
        return ans;
    }
};
